### PR TITLE
Work on master and worker to accommodate the new kind of gpu support

### LINKED ISSUE #459: on gpu-enabled clusters, kubernetes-master hooks sometimes thrash forever
See this [kubernetes-master debug-log output](https://gist.githubusercontent.com/Cynerva/90f4894e53ec9dacad28c01aa1625101/raw/6ceefde317d575da6c829b9c8ad71d7afdc63819/gistfile1.txt). A single hook gets stuck in a loop.

In the happy path, it goes like this:
1. `on_gpu_available` runs, setting `kubernetes-master.gpu.enabled` to true
2. `start_master` runs, sees `kubernetes-master.gpu.enabled` is set, and sets `kubernetes-master.privileged` to true
3. since `kubernetes-master.privileged` is true, `disable_gpu_mode` never runs.

In the failure case, `disable_gpu_mode` preempts start_master:
1. `on_gpu_available` runs, setting `kubernetes-master.gpu.enabled` to true
2. `disable_gpu_mode` runs, setting `kubernetes-master.gpu.enabled` to false
3. `start_master` runs, does not see `kubernetes-master.gpu.enabled`, so it sets `kubernetes-master.privileged` to false
4. `on_gpu_available` is eligible to run again, so... repeat

### PR DESCRIPTION
**Pull request by ktsakalozos**
**What this PR does / why we need it**: This PR adds support for the new kind of GPU/nvidia in Juju charms.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes: https://github.com/juju-solutions/bundle-canonical-kubernetes/issues/459

**Special notes for your reviewer**: This PR should go in with https://github.com/juju-solutions/layer-docker/pull/118

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py [hunk 1]
```diff
@@ -887,8 +891,9 @@ def on_gpu_available(kube_control):
     We need to run in privileged mode.
 
     """
+    kube_version = get_version('kube-apiserver')
     config = hookenv.config()
-    if config['allow-privileged'].lower() == "false":
+    if config['allow-privileged'] == "false" and kube_version < (1, 9):
```
Comment (Cynerva) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: Where'd the `.lower()` go? It was originally introduced in https://github.com/kubernetes/kubernetes/pull/59116 as a fix for https://github.com/juju-solutions/bundle-canonical-kubernetes/issues/374

File: cluster/juju/layers/kubernetes-master/config.yaml [hunk 2]
```diff
@@ -31,6 +31,11 @@ options:
       privileged mode. If "auto", kube-apiserver will not run in privileged
       mode by default, but will switch to privileged mode if gpu hardware is
       detected on a worker node.
+  automatic-nvidia-plugin-loading:
+    type: boolean
+    default: True
+    description: |
+      Automatically load the nvidia device plugin daemonset.
```
Comment (Cynerva) [inline cluster/juju/layers/kubernetes-master/config.yaml]: I'd feel safer with something like `enable-nvidia-plugin=auto`. Even if we only allow `false` or `auto` today, we might want to allow `true` in the future to override the automatic behavior. It'd be harder to grow into that with the boolean `automatic-nvidia-plugin-loading` config proposed here.

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py (hunk 1 above)
Comment (ktsakalozos) [reply cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: Probably missed it during a rebase :(

File: cluster/juju/layers/kubernetes-master/config.yaml (hunk 2 above)
Comment (ktsakalozos) [reply cluster/juju/layers/kubernetes-master/config.yaml]: Easy!

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py [hunk 3]
```diff
@@ -1102,14 +1122,6 @@ def configure_apiserver(etcd_connection_string, leader_etcd_version):
     server_cert_path = layer_options.get('server_certificate_path')
     server_key_path = layer_options.get('server_key_path')
 
-    # at one point in time, this code would set ca-client-cert,
-    # but this was removed. This was before configure_kubernetes_service
-    # kept track of old arguments and removed them, so client-ca-cert
-    # was able to hang around forever stored in the snap configuration.
-    # This removes that stale configuration from the snap if it still
-    # exists.
-    api_opts['client-ca-file'] = 'null'
```
Comment (Cynerva) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: Another casualty of the rebase?

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py [hunk 4]
```diff
@@ -1155,47 +1167,19 @@ def configure_apiserver(etcd_connection_string, leader_etcd_version):
...
-        api_opts['requestheader-allowed-names'] = 'client'
-        api_opts['requestheader-extra-headers-prefix'] = 'X-Remote-Extra-'
-        api_opts['requestheader-group-headers'] = 'X-Remote-Group'
-        api_opts['requestheader-username-headers'] = 'X-Remote-User'
-        api_opts['proxy-client-cert-file'] = client_cert_path
-        api_opts['proxy-client-key-file'] = client_key_path
-        api_opts['enable-aggregator-routing'] = 'true'
-        api_opts['client-ca-file'] = ca_cert_path
+        admission_control.remove('Initializers')
+    api_opts['admission-control'] = ','.join(admission_control)
```
Comment (Cynerva) [inline cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: A lot of accidental changes to api_opts and admission_control here, I think

File: cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py [hunk 5]
```diff
@@ -629,12 +630,10 @@ def configure_kubelet(dns, ingress_ip):
...
-        if get_version('kubelet') < (1, 6):
-            hookenv.log('Adding --experimental-nvidia-gpus=1 to kubelet')
-            kubelet_opts['experimental-nvidia-gpus'] = '1'
-        else:
-            hookenv.log('Adding --feature-gates=Accelerators=true to kubelet')
-            kubelet_opts['feature-gates'] = 'Accelerators=true'
+        hookenv.log('Adding '
+                    '--feature-gates=Accelerators=true,DevicePlugins=true '
+                    'to kubelet')
+        kubelet_opts['feature-gates'] = 'Accelerators=true,DevicePlugins=true'
```
Comment (Cynerva) [inline cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py]: Will this work with 1.7? According to the 1.7 kubelet docs here, DevicePlugins doesn't exist yet: https://v1-7.docs.kubernetes.io/docs/admin/kubelet/

Checking with @tvansteenburgh to see if we still need to support 1.7

Comment (Cynerva) [reply cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py]: Oh, nevermind, I see now that we only enable GPU support on 1.9+.

Comment (ktsakalozos) [reply cluster/juju/layers/kubernetes-worker/reactive/kubernetes_worker.py]: We do not install cuda anymore so any version less than 1.9 will not have gpu support.

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py (hunk 3 above)
Comment (ktsakalozos) [reply cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: Big fail...

File: cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py (hunk 4 above)
Comment (ktsakalozos) [reply cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py]: Major fail...

**Comment by Cynerva** [comment]
/lgtm

Thanks @ktsakalozos

**Pull request merged**